### PR TITLE
New field standards for `data_info`

### LINKED ISSUE #153: New `data_info` standard
Following discussions at the workshop, we need to implement a new `data_info` standard. Here is a proposal based on previous discussions. @thomasZen and @BaptisteLorenzi: please chime in if you have further input, I would start implementing the necessary changes to save these from next week.

First, I would propose to make the information also available under the alias `pretrained_cfg`, i.e. following similar conventions as [torch image models (`timm`)](https://huggingface.co/docs/timm/en/quickstart).

Then, I would propose saving the following fields in the dictionary:
- `n_neurons` : Dictionary, with sessions names as keys, and number of neurons in the sessions as values. This is already present.
- `mean`: tuple, with mean normalisation values for the videos. Would replace "movie_norm_dict" and "norm_mean" inside that, for ease of access.
- `std`: tuple, same as above. The idea is to flatten a bit the entries and make them more accessible
- `input_shape`: already present. Shape of the video input, excluding time dimension
- `response_rate`: float. Sampling rate of responses (i.e. the one used to train, so does not need to equate original experimental sampling rate for example after upsampling or downsampling from binning).
- `stimulus_rate`: float . Frame rate of the stimulus.
- `stimulus_class`: str, or list of str. The type of stimuli used to train the model, e.g. white noise, natural movies,... I do not see a need to standardise this right now, but we could settle on literal entries to use here. 
- `model_cut_frames`: int. Stimulus frames cut by the model (e.g. as a consequence of convolutions).
- `conv_type`: optional. Entry would be a string, indicating the convolution type the model is using.
- `core_type`: optional. Entry would be a string, indicating the type of core the model is using. 
- `readout_type`: optional. Entry would be a string, indicating which readout class is being used.
- `sessions_info`: optional. Nested dictionary with all extra session information that can be useful for downstream analyses. Keys would be session names, and values other dictionaries containing information either globally for the session (e.g. like the eye), or arrays containing information per neuron. This would be a potential candidate for standardisation.

I would vote for keeping the explainable variance per neuron / model performance of a neuron optional, and potentially going into sessions_info. I think it would still be useful to have an utility function that automatically adds at least the model performance at the end of training. I can add this in the same or a separate PR.

### PR DESCRIPTION
**Pull request by fededagos**
Fixes #153 . See issue for details.

### CODE REVIEW

**Comment by fededagos** [comment]
Right now this is working and tested only on hoefling_2024 models.

Outstanding TODOs: 

- [ ] Add pixel size and other metadata for datasets other than hoefling_2024
- [ ] Retrain and upload all models with updated data_info to huggingface

File: openretina/cli/train.py [hunk 1]
```diff
@@ -53,7 +53,7 @@ def train_model(cfg: DictConfig) -> float | None:
         movies_dictionary=movies_dict,
     )
 
-    data_info = compute_data_info(neuron_data_dict, movies_dict)
+    data_info = compute_data_info(neuron_data_dict, movies_dict, partial_data_info=cfg.data_io.get("data_info"))
```
Comment (thomasZen) [inline openretina/cli/train.py:56]: Does it work if partial_data_info is None? (`.get("data_info")` returns None if "data_info" is not in the dict)

File: openretina/data_io/base.py [hunk 2]
```diff
@@ -179,33 +179,58 @@ def normalize_train_test_movies(
...
+
+        input_shape = input_shapes[0]
+
     sessions_kwargs = {
         session_name: responses.session_kwargs for session_name, responses in neuron_data_dictionary.items()
     }
 
     return {
-        "n_neurons_dict": n_neurons_dict,
+        "n_neurons": n_neurons_dict,
```
Comment (thomasZen) [inline openretina/data_io/base.py]: any reason for this rename?

File: openretina/models/core_readout.py [hunk 3]
```diff
@@ -194,6 +220,11 @@ def update_model_data_info(self, data_info: dict[str, Any]) -> None:
             self.hparams["n_neurons_dict"] = self.data_info["n_neurons_dict"]
             self.hparams["data_info"] = self.data_info
 
+    @property
+    def pretrained_cfg(self) -> dict[str, Any]:
+        """Alias for data_info, following timm conventions."""
```
Comment (thomasZen) [inline openretina/models/core_readout.py]: maybe add a longer name or a link to the timm convention.
The first link that comes up on google is this one: https://www.timm2025.org/

File: openretina/models/core_readout.py [hunk 4]
```diff
@@ -194,6 +220,11 @@ def update_model_data_info(self, data_info: dict[str, Any]) -> None:
             self.hparams["n_neurons_dict"] = self.data_info["n_neurons_dict"]
```
Comment (thomasZen) [inline openretina/models/core_readout.py:220]: is this still up-to-date? There was a diff renaming "n_neurons_dict" to "n_neurons":
```
-        cfg.model.n_neurons_dict = data_info["n_neurons_dict"]
+        cfg.model.n_neurons_dict = data_info["n_neurons"]
```

File: openretina/modules/losses/correlation.py [hunk 5]
```diff
@@ -41,6 +51,11 @@ def forward(self, output, target, group_assignment, group_counts):
             (var_out + self.eps) * (var_target + self.eps)
         ).sqrt()
         corrs = corrs * (group_counts.sum() / group_counts.size(0) / group_counts[group_assignment - 1])[:, None, ...]
+
+        per_neuron_correlations = -corrs.view(-1, corrs.shape[-1]).mean(dim=0)
+        self._per_neuron_correlations = per_neuron_correlations.detach()
```
Comment (thomasZen) [inline openretina/modules/losses/correlation.py:56]: maybe add `.cpu().numpy()`, too. (I slightly prefer numpy over torch arrays for metadata like things)

Comment (thomasZen) [reply openretina/modules/losses/correlation.py:56]: but if doing though you'd have to adapt code below, too. Maybe just using `cpu()` or leaving as is is also good.

File: configs/data_io/hoefling_2024.yaml [hunk 6]
```diff
@@ -35,3 +35,9 @@ responses:
         _convert_: object
         file_path: ${paths.responses_path}
         cache_folder: ${paths.cache_dir}
+
+data_info:
+  response_rate: 30
+  stimulus_rate: 30
```
Comment (thomasZen) [inline configs/data_io/hoefling_2024.yaml]: if it's easy to change:
```suggestion
  response_rate_hz: 30
  stimulus_rate_hz: 30
```
as we mention units for the pixel size, too.

**Review by thomasZen — APPROVED** [review]
Looks good, added a few suggestions though.
Also it'd be good to triple check if you caught all n_neurons_dict to n_neurons renamings in case we stick to this decision.

File: openretina/cli/train.py (hunk 1 above)
Comment (fededagos) [reply openretina/cli/train.py:56]: Yes it will work, it was intentional :)

File: configs/data_io/hoefling_2024.yaml (hunk 6 above)
Comment (fededagos) [reply configs/data_io/hoefling_2024.yaml]: good point, will add

File: openretina/data_io/base.py (hunk 2 above)
Comment (fededagos) [reply openretina/data_io/base.py]: I thought it was a bit cleaner to have in the data info. But I think I will revert it. `n_neurons_dict` appears in too many places, and renaming it everywhere will likely introduce some problems...

File: openretina/models/core_readout.py (hunk 3 above)
Comment (fededagos) [reply openretina/models/core_readout.py]: It's pytorch image models, I have added it in parentheses now so it is clear.

File: openretina/models/core_readout.py [hunk 7]
```diff
@@ -104,16 +104,42 @@ def test_step(self, batch: tuple[str, DataPoint], batch_idx: int, dataloader_idx
         session_id, data_point = batch
         model_output = self.forward(data_point.inputs, session_id)
         loss = self.loss.forward(model_output, data_point.targets) / sum(model_output.shape)
-        correlation = -self.correlation_loss.forward(model_output, data_point.targets)
+        avg_correlation = -self.correlation_loss.forward(model_output, data_point.targets)
+        per_neuron_correlation = -self.correlation_loss._per_neuron_correlations
```
Comment (thomasZen) [inline openretina/models/core_readout.py]: This is a bit inconsistent for two reasons:
- we negate the per neuron correlations, and the name doesn't change (in the line above it goes from correlation_loss to avg_correlation after the negation which makes more sense)
- we access a private member

So maybe directly store the negated result in the `correlation_loss` class and make it public by calling it `per_neuron_correlations`.

Comment (fededagos) [reply openretina/models/core_readout.py]: I would want to keep it private as _per_neuron_correlations gets overwritten with every forward of the loss, storing the value of the last session essentially, so should only be accessed if people know what they are doing. (Also because if somebody wants explicitly the per_neuron loss only, the loss class offers an option to return only that already). I will resolve the naming inconsistency though

File: openretina/modules/losses/correlation.py (hunk 5 above)
Comment (fededagos) [reply openretina/modules/losses/correlation.py:56]: I would leave it as a tensor so it can stay a buffer of the module. But do you know if I can `.cpu()` it without problems? i.e. if the rest of the module is on `cuda` and I put that buffer on cpu, will something break?

Comment (thomasZen) [reply openretina/modules/losses/correlation.py:56]: > But do you know if I can .cpu() it without problems? i.e. if the rest of the module is on cuda and I put that buffer on cpu, will something break?

Okay, good point. It's probably better to keep all on the same device.
Older stackoverflow comments say you can have different parameters on different devices, but I think that's also confusing: https://stackoverflow.com/questions/58926054/how-to-get-the-device-type-of-a-pytorch-module-conveniently

File: openretina/data_io/base.py [hunk 8]
```diff
@@ -179,23 +179,49 @@ def normalize_train_test_movies(
...
+        stim_std = movies_dictionary.norm_std
+        input_shape = (
+            movies_dictionary.train.shape[0],
+            *movies_dictionary.train.shape[2:],
+        )
+    else:
+        norm_means = [movie.norm_mean for movie in movies_dictionary.values()]
+        norm_stds = [movie.norm_std for movie in movies_dictionary.values()]
+
+        if any(mean != norm_means[0] for mean in norm_means):
```
Comment (thomasZen) [inline openretina/data_io/base.py:210]: just as a not, it's also possible to write this check as:
```
if set(norm_means) > 1:
```
or to directly use a set above (but then indexing is harder an you'd have to write `next(iter(norm_means_set))` or so to get the value.

Anyway, just to show a potentially useful pattern, it's good as it is right now.

**Review by thomasZen — APPROVED** [review]
Looks good

**Comment by fededagos** [comment]
@thomasZen I am merging now. Before running training for all models I will open a separate refactoring PR, that I will use in the end for that.
